src/ddscxx/src/org/eclipse/cyclonedds/core/cdr/extended_cdr_v2_ser.cpp

```diff
@@ -217,7 +217,9 @@ const entity_properties_t* xcdr_v2_stream::first_entity(const entity_properties_
 bool xcdr_v2_stream::read_em_header(entity_properties_t &props)
 {
   uint32_t emheader = 0;
-  if (!read(*this,emheader))
+  size_t align_bytes = (4-(position()%4))%4;
```

Comment:
I was surprised by this, because it would seem likely that `read` takes care of aligning the stream and of checking whether sufficient data remains available. If I am not mistaken, it does. So something interesting that I'm not seeing must be going on here.

Reply:
The issue was that this needed to be done in two steps, first a "peek" to check whether there is enough space at the current stream level (as indicated by d-header) and ONLY if there is, then "take" the data from the stream into the emheader.

The old approach immediately tries to "take" the emheader from the stream, which might raise an error if there is not enough bytes left, which causes the streamer to enter the error state, and reject further deserialization.

So everything was working as intended, just a little bit too well.